ocaml/xapi/repository_helpers.ml

```diff
@@ -51,47 +123,104 @@ module Pkg = struct
+        error "%s" msg ;
+        raise Api_errors.(Server_error (internal_error, [msg]))
+
+  let to_fullname ~name ~arch ~epoch ~version ~release =
+    match epoch with
+    | Some i ->
+        Printf.sprintf "%s-%s:%s-%s.%s.rpm" name (string_of_int i) version
+          release arch
+    | None ->
+        Printf.sprintf "%s-%s-%s.%s.rpm" name version release arch
 end
 
 module Update = struct
   type t = {
       name: string
     ; arch: string
+    ; old_epoch: Epoch.t option
     ; old_version: string option
     ; old_release: string option
+    ; new_epoch: Epoch.t
     ; new_version: string
     ; new_release: string
     ; update_id: string option
     ; repository: string
   }
 
   let of_json j =
-    let open Yojson.Basic.Util in
-    {
-      name= member "name" j |> to_string
-    ; arch= member "arch" j |> to_string
-    ; new_version= member "newVerRel" j |> member "version" |> to_string
-    ; new_release= member "newVerRel" j |> member "release" |> to_string
-    ; old_version=
-        ( try Some (member "oldVerRel" j |> member "version" |> to_string)
-          with _ -> None
-        )
-    ; old_release=
-        ( try Some (member "oldVerRel" j |> member "release" |> to_string)
-          with _ -> None
-        )
-    ; update_id= member "updateId" j |> to_string_option
-    ; repository= member "repository" j |> to_string
-    }
+    try
+      let open Yojson.Basic.Util in
+      {
+        name= member "name" j |> to_string
+      ; arch= member "arch" j |> to_string
+      ; new_epoch=
+          member "newEpochVerRel" j
+          |> member "epoch"
+          |> to_string
+          |> Epoch.of_string
+      ; new_version= member "newEpochVerRel" j |> member "version" |> to_string
+      ; new_release= member "newEpochVerRel" j |> member "release" |> to_string
+      ; old_epoch=
+          ( try
+              Some
+                (member "oldEpochVerRel" j
+                |> member "epoch"
+                |> to_string
+                |> Epoch.of_string
+                )
+            with _ -> None
+          )
+      ; old_version=
+          ( try Some (member "oldEpochVerRel" j |> member "version" |> to_string)
+            with _ -> None
+          )
+      ; old_release=
+          ( try Some (member "oldEpochVerRel" j |> member "release" |> to_string)
+            with _ -> None
+          )
+      ; update_id= member "updateId" j |> to_string_option
+      ; repository= member "repository" j |> to_string
+      }
+    with e ->
+      let msg = "Can't construct an update from json" in
+      error "%s: %s" msg (ExnHelper.string_of_exn e) ;
+      raise Api_errors.(Server_error (internal_error, [msg]))
 
   let to_string u =
-    Printf.sprintf "%s.%s %s-%s -> %s-%s from %s:%s" u.name u.arch
+    Printf.sprintf "%s.%s %s:%s-%s -> %s:%s-%s from %s:%s" u.name u.arch
+      ( match u.old_epoch with
+      | Some e -> (
+        match e with Some i -> string_of_int i | None -> Epoch.epoch_none
```

Comment:
Is there a reason not to use `Epoch.to_string` here?

Reply:
Yes. Oh, silly me.